sos/report/plugins/juju.py

```diff
@@ -53,6 +82,71 @@ def setup(self):
             # logs in the directory.
             self.add_copy_spec("/var/log/juju/*.log")
 
+        # Only include the Juju state report if this plugin option is set
+        if not self.get_option("juju-state"):
+            return
+
+        juju_user = self.get_option("juju-user")
+        try:
+            pwd.getpwnam(juju_user)
+        except KeyError:
+            self._log_warn(
+                f'User "{juju_user}" does not exist, '
+                "will not collect Juju information."
+            )
+            return
+
+        if self.get_option("controllers") and self.get_option("models"):
+            self._log_warn(
+                "Options: controllers, models are mutually exclusive. "
+                "Will not collect Juju information."
+            )
+            return
+
+        controllers_json = self.collect_cmd_output(
+            "juju controllers --format=json", runas=juju_user
+        )
+        if controllers_json["status"] == 0:
+            desired_controllers = set(
+                self.get_option("controllers").split(" ")
+            )
+            # If a controller option is supplied, use it. Otherwise, get all
+            #  controllers
+            if desired_controllers and desired_controllers != {""}:
+                controllers = desired_controllers
+            else:
+                controllers = set(
+                    json.loads(controllers_json["output"])[
+                        "controllers"
+                    ].keys()
+                )
+        else:
+            controllers = {}
+
+        # Specific models
+        if self.get_option("models"):
+            for model in self.get_option("models").split(" "):
+                command = f"juju status -m {model} --format=json"
+                self.collect_cmd_output(command, runas=juju_user)
+
+        # All controllers and all models OR specific controllers and all
+        #  models for each
+        else:
+            for controller in controllers:
+                models_json = self.exec_cmd(
+                    f"juju models --all -c {controller} --format=json",
+                    runas=juju_user,
+                )
+                if models_json["status"] == 0:
+                    models = json.loads(models_json["output"])["models"]
+                    for model in models:
+                        short_name = model["short-name"]
+                        command = (
+                            f"juju status -m {controller}:{short_name} "
+                            f"--format=json"
+                        )
+                        self.collect_cmd_output(command, runas=juju_user)
```

Comment:
Same here, use `add_cmd_output()` if you don't need to do anything further with the output.